### ISSUE TITLE
Locate classes directly via namespace, rather than relying on filename conventions

### ISSUE THREAD

**Issue by alexweissman** [userfrosting/UserFrosting#743]
There are a couple places where this comes up:

- Service provider classes for Sprinkles
- Database migration files
- Sprinkle bootstrapper classes

In all of these cases, we rely on a naming convention - specifically, that the StudlyCaps™️ version of the Sprinkle name will correspond to the namespace of the service provider/migration/whatever.  For example, `extend-user` -> `ExtendUser`.

The problem is that a lot of developers (especially newbie devs) struggle with proper naming conventions, or simply don't want to use this convention.

Unfortunately, PHP [doesn't provide a way](https://stackoverflow.com/questions/22761554/php-get-all-class-names-inside-a-particular-namespace) to "glob" a namespace in the same way that you can glob files.  There are some packages that can do this, e.g. https://github.com/gnugat/nomo-spaco, but I'm not sure how well-maintained they are or what their performance is like.  In the case of migrations, it's not so important, but with service providers and Sprinkle bootstrappers, we'd probably want to cache this namespace tree.

Frankly, I wouldn't be surprised it Composer itself provided some sort of API for this.  After all, Composer is already going through and scanning every file to construct the mappings of filenames -> class names.

**Comment by alexweissman** [comment]
I think we could also do autoloading for Twig extensions as well.

**Comment by alexweissman** [comment]
This will help a lot, in conjunction with #729, for people who are struggling with the naming conventions at the beginning.

**Comment by elliottpost** [comment]
Another use case for this: if my sprinkle directory name is `card-manager` and my PSR-4 autoloader in `composer.json` reads: 

```javascript 
{
...
    "psr-4": {
      "UserFrosting\\Sprinkle\\Card\\Manager\\": "src/"
    }
...
}
```

instead of


```javascript 
{
...
    "psr-4": {
      "UserFrosting\\Sprinkle\\CardManager\\": "src/"
    }
...
}
```

UF will fail to correctly locate the DB migrations, even though they are mapped correctly in composer's autoload files.

TL;DR: UF relying on file convention instead of namespace requires a very strict naming convention, even if it agrees to PSR-4 standards. Until the issue can be addressed, it's probably worth a tip in the docs that UF does not support Sprinkle namespaces beyond one level.

**Comment by lcharette** [comment]
FYI, the DB migration convention is a known issue. It's in part because PHP doesn't have the ability to list classes (or something like that). I should probably open an issue for that...

That said, a sprinkle name beyond one level doesn't really make sense. In the `UserFrosting\\Sprinkle\\Card\\Manager\\` example, you can have a sprinkle with the `Card` name and `Manager` can be under it. There's no real logic as to why a sprinkle can have a sub sprinkle (and where to place it in the doc structure). The only way I see this as beneficial would be to include the vendor name in the Sprinkle namespace. To support this, we would need to have had a `UserFrosting\\Sprinkle\\vendor\\SpinkleName` convention (breaking support for older versions).

**Comment by elliottpost** [comment]
I disagree on the bit about the sprinkle name beyond one level, here's why: if Sprinkles are like 'plugins', then it makes sense to make small, modular sprinkles. So, if there's multiple aspects to `Card` in this case, the idea is to have a sprinkle for Card Management, and perhaps another sprinkle for `Card API`, or `Card Interactions`, etc. Sure we could put it all under one large sprinkle, but that seems counter-intuitive, to me, of having the idea of small, modular, sprinkles that can be dropped-in to toggle additional functionality on/off. 

Instead then of having `UserFrosting\Sprinkle\Card\API\...` and `UserFrosting\Sprinkle\Card\Manager\`, etc, we have them all under their own namespace. It's a trivial issue, but it seems to be that if they have a relationship, they should be within a shared namespace, even if it's more logical to separate them by sprinkle based on functionality. Especially if one is not dependent upon another. 

My point, in short, is that you can get better loose-coupling by not requiring that structure.

**Comment by lcharette** [comment]
Hum... True. Something `Vendor` name could indeed solve...

**Comment by elliottpost** [comment]
Well, it's a bit of an abstract usage case that may not affect many others anyway, but wanted to chime in with my 2 cents since it came up for me.

**Comment by lcharette** [comment]
This is solved in V5, with Sprinkle Recipe
